src/VisualStudio/Core/Def/Implementation/EditAndContinue/DebuggeeModuleMetadataProvider.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Debugging;
+using Microsoft.CodeAnalysis.EditAndContinue;
+using Microsoft.VisualStudio.Debugger;
+using Microsoft.VisualStudio.Debugger.Clr;
+using Microsoft.VisualStudio.Debugger.ComponentInterfaces;
+
+namespace Microsoft.VisualStudio.LanguageServices.EditAndContinue
+{
+    [Export(typeof(IDebuggeeModuleMetadataProvider)), Shared]
+    internal sealed class DebuggeeModuleMetadataProvider : IDebuggeeModuleMetadataProvider
+    {
+        /// <summary>
+        /// Concord component. Singleton created on demand during debugging session and discarded as soon as the session ends.
+        /// </summary>
+        private sealed class DebuggerService : IDkmCustomMessageForwardReceiver
+        {
+            /// <summary>
+            /// Component id as specified in ManagedEditAndContinueService.vsdconfigxml.
+            /// </summary>
+            public static readonly Guid ComponentId = new Guid("A96BBE03-0408-41E3-8613-6086FD494B43");
+
+            /// <summary>
+            /// Message source id as specified in ManagedEditAndContinueService.vsdconfigxml.
+            /// </summary>
+            public static readonly Guid MessageSourceId = new Guid("58CDF976-1923-48F7-8288-B4189F5700B1");
+
+            private sealed class DataItem : DkmDataItem
+            {
+                private readonly DebuggeeModuleMetadataProvider _provider;
+                private readonly Guid _mvid;
+
+                public DataItem(DebuggeeModuleMetadataProvider provider, Guid mvid)
+                {
+                    _provider = provider;
+                    _mvid = mvid;
+                }
+
+                protected override void OnClose() => _provider.OnModuleInstanceUnload(_mvid);
+            }
+
+            DkmCustomMessage IDkmCustomMessageForwardReceiver.SendLower(DkmCustomMessage customMessage)
+            {
+                var provider = (DebuggeeModuleMetadataProvider)customMessage.Parameter1;
+                var clrModuleInstance = (DkmClrModuleInstance)customMessage.Parameter2;
+
+                // Note that this call has to be made in a Concord component.
+                // The debugger tracks what the current component is and associates the data item with it.
+                clrModuleInstance.SetDataItem(DkmDataCreationDisposition.CreateAlways, new DataItem(provider, clrModuleInstance.Mvid));
+                return null;
+            }
+        }
+
+        private readonly DebuggeeModuleMetadataCache _baselineMetadata;
+
+        public DebuggeeModuleMetadataProvider()
+        {
+            _baselineMetadata = new DebuggeeModuleMetadataCache();
+        }
+
+        private void OnModuleInstanceUnload(Guid mvid)
+            => _baselineMetadata.Remove(mvid);
```

Comment:
Should the `ModuleMetadata` instances also need to be somehow disposed in case another thread has obtained one at the time the module instance is unloaded?